testing/test_pythonpath.py

```diff
@@ -0,0 +1,81 @@
+import sys
+from textwrap import dedent
+
+import pytest
+from _pytest.pytester import Pytester
+
+
+@pytest.fixture()
+def file_structure(pytester: Pytester) -> None:
+    pytester.makepyfile(
+        test_foo="""
+        from foo import foo
+
+        def test_foo():
+            assert foo() == 1
+        """
+    )
+
+    pytester.makepyfile(
+        test_bar="""
+        from bar import bar
+
+        def test_bar():
+            assert bar() == 2
+        """
+    )
+
+    foo_py = pytester.mkdir("sub") / "foo.py"
+    content = dedent(
+        """
+        def foo():
+            return 1
+        """
+    )
+    foo_py.write_text(content, encoding="utf-8")
+
+    bar_py = pytester.mkdir("sub2") / "bar.py"
+    content = dedent(
+        """
+        def bar():
+            return 2
+        """
+    )
+    bar_py.write_text(content, encoding="utf-8")
+
+
+def test_one_dir(pytester: Pytester, file_structure) -> None:
+    pytester.makefile(".ini", pytest="[pytest]\npythonpath=sub\n")
+    result = pytester.runpytest("test_foo.py")
+    result.assert_outcomes(passed=1)
+
+
+def test_two_dirs(pytester: Pytester, file_structure) -> None:
+    pytester.makefile(".ini", pytest="[pytest]\npythonpath=sub sub2\n")
+    result = pytester.runpytest("test_foo.py", "test_bar.py")
+    result.assert_outcomes(passed=2)
+
+
+def test_module_not_found(pytester: Pytester, file_structure) -> None:
+    """If pythonpath setting not there, test should error."""
+    pytester.makefile(".ini", pytest="[pytest]\n")
+    result = pytester.runpytest("test_foo.py")
+    result.assert_outcomes(errors=1)
+    expected_error = "E   ModuleNotFoundError: No module named 'foo'"
+    result.stdout.fnmatch_lines([expected_error])
+
+
+def test_no_ini(pytester: Pytester, file_structure) -> None:
+    """If no ini file, test should error."""
+    result = pytester.runpytest("test_foo.py")
+    result.assert_outcomes(errors=1)
+    expected_error = "E   ModuleNotFoundError: No module named 'foo'"
+    result.stdout.fnmatch_lines([expected_error])
+
+
+def test_path_removal(pytester: Pytester, file_structure) -> None:
+    """Make sure sys.path is cleaned up after testing."""
+    original = sys.path.copy()
+    pytester.makefile(".ini", pytest="[pytest]\npythonpath=sub sub2\n")
+    pytester.runpytest()
+    assert sys.path == original
```

Comment:
@bluetech I attempted to add cleanup code, and things blew up. So I added this test `test_path_removal()`.
I'm not sure where in pytest the path is being restored, but it looks like it's already done somewhere. 
That makes sense actually, as pytest needs to add the paths for test directories, so cleanup is already necessary.

Reply:
Ok. I'll have to play around a bit to figure out a non-pytester way to test this.

Reply:
I re-added the cleanup code. My problem before was that I was trying to remove path objects from sys.path instead of removing  strings. Fixed now. Also ran some tests on the side to convince myself that you were correct about the paths getting left around.

I went with a simple approach of:
1. always adding each path to sys.path, even if it's already there, at the front of the list.
2. always removing each added path using remove(). 

Since `list.remove(x)` only removes the first instance of x, I think we're fine.
Even if a path already exists in sys.path, we know we added it to the front, so removing it from the front seems like it should be safe.